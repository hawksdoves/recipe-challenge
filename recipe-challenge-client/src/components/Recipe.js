import React from 'react';
import { Card } from 'semantic-ui-react';

import Ingredient from './Ingredient';

function Recipe(props) {
    const { recipe, children } = props;

    return recipe ? (
        <Card key={recipe.id}>
            <Card.Content>
                {children}
                <Card.Header style={{margin: 10}}>{recipe.name}</Card.Header>
                <Card.Meta>{recipe.time} minutes</Card.Meta>
                {recipe.ingredients.map(ingredient => {
                    return (
                        <Ingredient ingredient={ingredient} key={`${recipe.id}-${ingredient.name}`}/>
                    )
                })}
            </Card.Content>
        </Card>
    ) :
    null
}

export default Recipe;
