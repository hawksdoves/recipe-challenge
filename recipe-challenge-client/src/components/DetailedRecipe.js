import React from 'react';
import { Link } from 'react-router-dom';

import Recipe from './Recipe';
import Favourite from './Favourite';
import getRecipe from '../services/getRecipe';

class DetailedRecipe extends React.Component {

    constructor(props) {
        super(props);
        this.state = { recipe: undefined };
    }

    componentDidMount() {
        const { pathname } = this.props.location;
        const id = pathname.substr(-1);
        getRecipe(id)
            .then(resp => this.setState({ recipe: resp.data }));
    }

    render() {
        const { favourites, ...rest } = this.props;

        return  (
            <div className='detailedRecipe'>
                <Link to='/'>Return to list</Link>
                { this.state.recipe ? (
                    <Recipe recipe={this.state.recipe}>
                        <Favourite favourite={favourites.includes(this.state.recipe.id)} {...rest} recipeId={this.state.recipe.id} />
                    </Recipe>
                    ) :
                    <div>Sorry, this recipe doesn't exist or may have been removed</div>
                }    
            </div>
        )  
    } 
};

export default DetailedRecipe;