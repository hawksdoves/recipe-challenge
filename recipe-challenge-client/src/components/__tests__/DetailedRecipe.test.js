import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import Recipe from '../Recipe';

const mockResponse = {
    data: {
        "id": "1",
        "name": "Lemon Chicken",
        "time": "30",
        "ingredients": [
            {
                "name": "Chicken",
                "quantity": "4"
            },
            {
                "name": "Lemon",
                "quantity": "1"
            },
            {
                "name": "Thyme",
                "quantity": "1",
                "measure": "tsp"
            }
        ]
    }
}

const API_URL = 'http://localhost:3001/api';

jest.mock('axios', () => {
    const mockGetRequest = jest.fn(() => Promise.resolve(mockResponse));

    return {
        get: mockGetRequest
    };
});

import DetailedRecipe from '../DetailedRecipe';

describe('DetailedRecipe', () => {
    it('should render the recipe correctly', () =>{
        const wrapper = shallow(<DetailedRecipe favourites={[]}/>);
        wrapper.setState({ recipe: mockResponse.data });
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    })

    it('should display message when the recipe does not exist', () =>{
        const wrapper = shallow(<DetailedRecipe favourites={[]}/>);
        expect(wrapper).toMatchSnapshot();
    })

    it('calls componentDidMount() lifecycle method', () => {
        const mockParamsId = { pathname: 'd/1' };
        const componentDidMountSpy = jest.spyOn(DetailedRecipe.prototype, 'componentDidMount');
        const wrapper = mount(<MemoryRouter><DetailedRecipe location={mockParamsId} favourites={[]}/></MemoryRouter>);
        expect(componentDidMountSpy).toHaveBeenCalled();
    });
});
