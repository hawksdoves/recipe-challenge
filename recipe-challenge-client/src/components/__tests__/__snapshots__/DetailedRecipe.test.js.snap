// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedRecipe should display message when the recipe does not exist 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="detailedRecipe"
>
    <Link
        replace={false}
        to="/"
    >
        Return to list
    </Link>
    <div>
        Sorry, this recipe doesn't exist or may have been removed
    </div>
</div>,
  "nodes": Array [
    <div
      className="detailedRecipe"
>
      <Link
            replace={false}
            to="/"
      >
            Return to list
      </Link>
      <div>
            Sorry, this recipe doesn't exist or may have been removed
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DetailedRecipe
        favourites={Array []}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DetailedRecipe {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "favourites": Array [],
        },
        "refs": Object {},
        "state": Object {
          "recipe": undefined,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="detailedRecipe"
>
          <Link
                    replace={false}
                    to="/"
          >
                    Return to list
          </Link>
          <div>
                    Sorry, this recipe doesn't exist or may have been removed
          </div>
</div>,
        "_debugID": 4,
        "_renderedOutput": <div
          className="detailedRecipe"
>
          <Link
                    replace={false}
                    to="/"
          >
                    Return to list
          </Link>
          <div>
                    Sorry, this recipe doesn't exist or may have been removed
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DetailedRecipe
    favourites={Array []}
/>,
}
`;

exports[`DetailedRecipe should render the recipe correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="detailedRecipe"
>
    <Link
        replace={false}
        to="/"
    >
        Return to list
    </Link>
    <Recipe
        recipe={
            Object {
                "id": "1",
                "ingredients": Array [
                  Object {
                    "name": "Chicken",
                    "quantity": "4",
                  },
                  Object {
                    "name": "Lemon",
                    "quantity": "1",
                  },
                  Object {
                    "measure": "tsp",
                    "name": "Thyme",
                    "quantity": "1",
                  },
                ],
                "name": "Lemon Chicken",
                "time": "30",
              }
        }
    >
        <Favourite
            favourite={false}
            recipeId="1"
        />
    </Recipe>
</div>,
  "nodes": Array [
    <div
      className="detailedRecipe"
>
      <Link
            replace={false}
            to="/"
      >
            Return to list
      </Link>
      <Recipe
            recipe={
                  Object {
                        "id": "1",
                        "ingredients": Array [
                          Object {
                            "name": "Chicken",
                            "quantity": "4",
                          },
                          Object {
                            "name": "Lemon",
                            "quantity": "1",
                          },
                          Object {
                            "measure": "tsp",
                            "name": "Thyme",
                            "quantity": "1",
                          },
                        ],
                        "name": "Lemon Chicken",
                        "time": "30",
                      }
            }
      >
            <Favourite
                  favourite={false}
                  recipeId="1"
            />
      </Recipe>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DetailedRecipe
        favourites={Array []}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DetailedRecipe {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "favourites": Array [],
        },
        "refs": Object {},
        "state": Object {
          "recipe": Object {
            "id": "1",
            "ingredients": Array [
              Object {
                "name": "Chicken",
                "quantity": "4",
              },
              Object {
                "name": "Lemon",
                "quantity": "1",
              },
              Object {
                "measure": "tsp",
                "name": "Thyme",
                "quantity": "1",
              },
            ],
            "name": "Lemon Chicken",
            "time": "30",
          },
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="detailedRecipe"
>
          <Link
                    replace={false}
                    to="/"
          >
                    Return to list
          </Link>
          <Recipe
                    recipe={
                              Object {
                                        "id": "1",
                                        "ingredients": Array [
                                          Object {
                                            "name": "Chicken",
                                            "quantity": "4",
                                          },
                                          Object {
                                            "name": "Lemon",
                                            "quantity": "1",
                                          },
                                          Object {
                                            "measure": "tsp",
                                            "name": "Thyme",
                                            "quantity": "1",
                                          },
                                        ],
                                        "name": "Lemon Chicken",
                                        "time": "30",
                                      }
                    }
          >
                    <Favourite
                              favourite={false}
                              recipeId="1"
                    />
          </Recipe>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="detailedRecipe"
>
          <Link
                    replace={false}
                    to="/"
          >
                    Return to list
          </Link>
          <Recipe
                    recipe={
                              Object {
                                        "id": "1",
                                        "ingredients": Array [
                                          Object {
                                            "name": "Chicken",
                                            "quantity": "4",
                                          },
                                          Object {
                                            "name": "Lemon",
                                            "quantity": "1",
                                          },
                                          Object {
                                            "measure": "tsp",
                                            "name": "Thyme",
                                            "quantity": "1",
                                          },
                                        ],
                                        "name": "Lemon Chicken",
                                        "time": "30",
                                      }
                    }
          >
                    <Favourite
                              favourite={false}
                              recipeId="1"
                    />
          </Recipe>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DetailedRecipe
    favourites={Array []}
/>,
}
`;
